### ISSUE TITLE
Same value for Accuracy, Precision and Recall for MNIST classification with 10 classes

### ISSUE THREAD

**Issue by lsisoft** [PyTorchLightning/metrics#298]
## 🐛 Bug

When computing torchmetrics Accuracy, Precision, Recall and F1 over MNIST classification, all numbers come up the same.

Also, in this case, torchmetrics Accuracy metric sets the mode to multiclass, not multilabel, so it uses exactly the same formula as Precision.

I've attached a reduced code sample below, and a larger one is available on: https://gist.github.com/lsisoft/e4933683131872976a00be3ae0fc88ea - where I have attempted to rewrite the metrics as inline versions, forcing Accuracy to use the proper formula.

### To Reproduce

1. Run the code sample below
2. Monitor metrics in tensorboard or by debugging in the code

#### Code sample

```python
import pytorch_lightning as pl
import torch
import torchmetrics
import torchmetrics.functional
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torch.utils.data import random_split
from torchvision import transforms
from torchvision.datasets import MNIST


class LitCritic(pl.LightningModule):
    def __init__(self):
        super().__init__()

        self.encoder = nn.Sequential(
            nn.Linear(28 * 28, 64),
            nn.ReLU(),
            nn.Linear(64, 64),
            nn.ReLU(),
            nn.Linear(64, 10))

        metrics = torchmetrics.MetricCollection([
            # Accuracy: due to mode multiclass, not multilabel, this uses same formula as Precision
            torchmetrics.Accuracy(num_classes=10),
            torchmetrics.Precision(num_classes=10),
            torchmetrics.Recall(num_classes=10),
            torchmetrics.F1(num_classes=10),
        ])

        self.train_metrics = metrics.clone(prefix='train_')
        self.valid_metrics = metrics.clone(prefix='val_')

    def forward(self, x):
        embedding = self.encoder(x)
        return embedding

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

    def training_step(self, train_batch, batch_idx):
        x, y = train_batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)

        loss = F.cross_entropy(z, y)
        self.log('train_loss', loss)

        z_prob = F.softmax(z, dim=-1)
        output = self.train_metrics(z_prob, y)
        self.log_dict(output)

        return loss

    def validation_step(self, val_batch, batch_idx):
        x, y = val_batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)

        loss = F.cross_entropy(z, y)
        self.log('val_loss', loss)

        z_prob = F.softmax(z, dim=-1)
        output = self.valid_metrics(z_prob, y)
        self.log_dict(output)


if __name__ == '__main__':
    # data
    dataset = MNIST('', train=True, download=True, transform=transforms.ToTensor())
    mnist_train, mnist_val = random_split(dataset, [55000, 5000])

    batch_size = 32
    train_loader = DataLoader(mnist_train, batch_size=batch_size)
    val_loader = DataLoader(mnist_val, batch_size=batch_size)

    # model
    model = LitCritic()

    # training
    trainer = pl.Trainer()
    trainer.fit(model, train_loader, val_loader)
```

### Expected behavior

Each metric should return a different value.

### Environment

 - PyTorch Version (e.g., 1.0): 1.8.1+cu111
 - OS (e.g., Linux): Ubuntu
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.8.10
 - CUDA/cuDNN version: CUDA 11.3
 - GPU models and configuration: GTX 1080
 - Any other relevant information:

### Additional context

None

**Comment by SkafteNicki** [comment]
Hi @lsisoft , could you please provide which version of torchmetrics you are using?

**Comment by lsisoft** [comment]
torchmetrics==0.3.2

**Comment by lsisoft** [comment]
I have reviewed your code snippet and indeed it matches sklearn. I've moved my original code to macro and now the values for all metrics are almost what they should be.

There is a an issue that remains: Accuracy vs Precision.

For Accuracy, torchmetrics/functional/classification/accuracy.py, if I can place a debug breakpoint at line 80, it hits, due to mode=DataType.MULTICLASS

The code there looks like this:
    if (mode == DataType.BINARY and average in simple_average) or mode == DataType.MULTILABEL:
        numerator = tp + tn
        denominator = tp + tn + fp + fn
    else:
        numerator = tp # <-- breakpoint
        denominator = tp + fn

and due to going on the else branch, it uses same formula for Accuracy as for precision.

**Comment by SkafteNicki** [comment]
@lsisoft could you give me an example where the calculated value in torchmetrics deviates from the value in sklearn?

**Comment by lsisoft** [comment]
If your criteria is that it has to only match sklearn, than you can close the issue. To my is strange that it uses the wrong formula in torchmetrics and sklearn as well.

**Comment by SkafteNicki** [comment]
Yes our criteria is to match sklearn because they are the trusted standard for calculating these kind of metrics.
That said, I looked more into this and there is actually is no problem with that formula. The typical `(tp + tn) / (tp + tn + fp + fn)` is *ONLY* true when we are evaluating binary and multilabel examples. In the multiclass case we basically just need to compute `correct/total` which is exactly what our code does. Closing this issue.

**Comment by randombenj** [comment]
I don't quite get why this should be correct (both in sklearn and torchmetrics). But maybe I just don't understand the reasoning behind this.  This also led to some confusion in a peer review.
Given the following confusion matrix:

```
[[105   0   0   4   3]
 [  0  95   0  15   3]
 [  0   0   0   2   0]
 [  1   0   0   2   0]
 [  1   0   0   9   2]]
```
    
 An example for Class 0 (index 0), given `TP = 105`, `TN = 95 + 15 + 3 + 2 + 2 + 9 + 2 = 128`, `FP = 2`, `FN = 7`.

Accuracy = `(TN + TP) / (N + P) = (128 + 105) / (128 + 7 + 105 + 2) = 0.9628`
Recall/Sensitivity = `TP / (TN + FN) = 105 / (105 + 7) = 0.9375`


So when I calculate:

```python
from torchmetrics.classification import Accuracy
from torch.nn.functional import one_hot

NUM_CLASSES = 7
probability_batch 

# -- current implementation (multiclass) --

_target = torch.tensor([y for (_, y) in probability_batch])
_preds = torch.tensor([torch.argmax(p).cpu() for (p, _) in probability_batch])

print("_preds", _preds)
# _preds tensor([0, 0, 0, 5, 0, 0, 3, 0, 0, 0, 1, 1, 1, 3, 3, 1, 1, 1, 0, 0, 1, 3, 1, 1,
#         1, 1, 1, 1, 0, 0, 0, 0, 0, 3, 5, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 5,
#         0, 0, 1, 3, 1, 3, 1, 0, 1, 1, 0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 0, 0, 3, 0,
#         0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 3, 1, 1, 0, 0, 1,
#         1, 3, 0, 0, 3, 1, 1, 5, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 5, 0, 0, 0,
#         1, 1, 0, 1, 1, 0, 5, 0, 0, 0, 0, 3, 3, 0, 3, 0, 3, 5, 0, 0, 0, 1, 0, 3,
#         0, 0, 0, 1, 1, 0, 3, 1, 1, 1, 0, 1, 1, 1, 3, 1, 0, 0, 0, 0, 0, 1, 1, 3,
#         0, 1, 1, 0, 0, 3, 1, 1, 1, 1, 0, 5, 0, 0, 3, 0, 1, 1, 1, 1, 3, 3, 0, 1,
#         1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 3, 0, 1, 1, 1, 1, 3, 0, 1, 1, 0, 1, 1, 0,
#         3, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 3, 3, 1, 0, 0, 0, 1, 3,
#         0, 3])

print("_target", _target)
# _preds tensor([0, 0, 0, 5, 0, 0, 3, 0, 0, 0, 1, 1, 1, 3, 3, 1, 1, 1, 0, 0, 1, 3, 1, 1,
#         1, 1, 1, 1, 0, 0, 0, 0, 0, 3, 5, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 5,
#         0, 0, 1, 3, 1, 3, 1, 0, 1, 1, 0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 0, 0, 3, 0,
#         0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 3, 1, 1, 0, 0, 1,
#         1, 3, 0, 0, 3, 1, 1, 5, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 5, 0, 0, 0,
#         1, 1, 0, 1, 1, 0, 5, 0, 0, 0, 0, 3, 3, 0, 3, 0, 3, 5, 0, 0, 0, 1, 0, 3,
#         0, 0, 0, 1, 1, 0, 3, 1, 1, 1, 0, 1, 1, 1, 3, 1, 0, 0, 0, 0, 0, 1, 1, 3,
#         0, 1, 1, 0, 0, 3, 1, 1, 1, 1, 0, 5, 0, 0, 3, 0, 1, 1, 1, 1, 3, 3, 0, 1,
#         1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 3, 0, 1, 1, 1, 1, 3, 0, 1, 1, 0, 1, 1, 0,
#         3, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 3, 3, 1, 0, 0, 0, 1, 3,
#         0, 3])



acc = Accuracy(task="multiclass", num_classes=NUM_CLASSES, labels=dataset.classes, average='weighted')
# tensor(0.8430)

acc = Accuracy(task="multiclass", num_classes=NUM_CLASSES, labels=dataset.classes, average=None)
# tensor([0.9813, 1.0000, 0.0000, 0.0625, 0.0000, 0.2500, 0.0000])

# -- expected implementation (according to Accuracy = `(TN + TP) / (N + P) = (128 + 105) / (128 + 7 + 105 + 2)`)


_target = torch.tensor([one_hot(torch.tensor(y), num_classes=7).tolist() for y in _target])
_preds = torch.tensor([one_hot(torch.argmax(p).cpu(), num_classes=7).tolist() for p) in _preds])

acc = Accuracy(task="multilabel", num_labels=NUM_CLASSES, labels=dataset.classes, average='weighted')
# tensor(0.9352)

acc = Accuracy(task="multilabel", num_labels=NUM_CLASSES, labels=dataset.classes, average=None)
# tensor([0.9628, 0.9256, 0.9917, 0.8719, 1.0000, 0.9339, 1.0000])
```

Of course the behaviour between sklearn and torchmetrics is constant, so this is probably me ...
Just for anyone else stumpling upon this issue :sweat_smile:

**Comment by randombenj** [comment]
I don't quite get why this should be correct (both in sklearn and torchmetrics). But maybe I just don't understand the reasoning behind this.  This also led to some confusion in a peer review.
Given the following confusion matrix:

```
[[105   0   0   4   3]
 [  0  95   0  15   3]
 [  0   0   0   2   0]
 [  1   0   0   2   0]
 [  1   0   0   9   2]]
```
    
 An example for Class 0 (index 0), given `TP = 105`, `TN = 95 + 15 + 3 + 2 + 2 + 9 + 2 = 128`, `FP = 2`, `FN = 7`.

Accuracy = `(TN + TP) / (N + P) = (128 + 105) / (128 + 7 + 105 + 2) = 0.9628`
Recall/Sensitivity = `TP / (TN + FN) = 105 / (105 + 7) = 0.9375`


So when I calculate:

```python
from torchmetrics.classification import Accuracy
from torch.nn.functional import one_hot

NUM_CLASSES = 7
probability_batch 

# -- current implementation (multiclass) --

_target = torch.tensor([y for (_, y) in probability_batch])
_preds = torch.tensor([torch.argmax(p).cpu() for (p, _) in probability_batch])

print("_preds", _preds)
# _preds tensor([0, 0, 0, 5, 0, 0, 3, 0, 0, 0, 1, 1, 1, 3, 3, 1, 1, 1, 0, 0, 1, 3, 1, 1,
#         1, 1, 1, 1, 0, 0, 0, 0, 0, 3, 5, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 5,
#         0, 0, 1, 3, 1, 3, 1, 0, 1, 1, 0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 0, 0, 3, 0,
#         0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 3, 1, 1, 0, 0, 1,
#         1, 3, 0, 0, 3, 1, 1, 5, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 5, 0, 0, 0,
#         1, 1, 0, 1, 1, 0, 5, 0, 0, 0, 0, 3, 3, 0, 3, 0, 3, 5, 0, 0, 0, 1, 0, 3,
#         0, 0, 0, 1, 1, 0, 3, 1, 1, 1, 0, 1, 1, 1, 3, 1, 0, 0, 0, 0, 0, 1, 1, 3,
#         0, 1, 1, 0, 0, 3, 1, 1, 1, 1, 0, 5, 0, 0, 3, 0, 1, 1, 1, 1, 3, 3, 0, 1,
#         1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 3, 0, 1, 1, 1, 1, 3, 0, 1, 1, 0, 1, 1, 0,
#         3, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 3, 3, 1, 0, 0, 0, 1, 3,
#         0, 3])

print("_target", _target)
# _preds tensor([0, 0, 0, 5, 0, 0, 3, 0, 0, 0, 1, 1, 1, 3, 3, 1, 1, 1, 0, 0, 1, 3, 1, 1,
#         1, 1, 1, 1, 0, 0, 0, 0, 0, 3, 5, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 5,
#         0, 0, 1, 3, 1, 3, 1, 0, 1, 1, 0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 0, 0, 3, 0,
#         0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 3, 1, 1, 0, 0, 1,
#         1, 3, 0, 0, 3, 1, 1, 5, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 5, 0, 0, 0,
#         1, 1, 0, 1, 1, 0, 5, 0, 0, 0, 0, 3, 3, 0, 3, 0, 3, 5, 0, 0, 0, 1, 0, 3,
#         0, 0, 0, 1, 1, 0, 3, 1, 1, 1, 0, 1, 1, 1, 3, 1, 0, 0, 0, 0, 0, 1, 1, 3,
#         0, 1, 1, 0, 0, 3, 1, 1, 1, 1, 0, 5, 0, 0, 3, 0, 1, 1, 1, 1, 3, 3, 0, 1,
#         1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 3, 0, 1, 1, 1, 1, 3, 0, 1, 1, 0, 1, 1, 0,
#         3, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 3, 3, 1, 0, 0, 0, 1, 3,
#         0, 3])



acc = Accuracy(task="multiclass", num_classes=NUM_CLASSES, labels=dataset.classes, average='weighted')
# tensor(0.8430)

acc = Accuracy(task="multiclass", num_classes=NUM_CLASSES, labels=dataset.classes, average=None)
# tensor([0.9813, 1.0000, 0.0000, 0.0625, 0.0000, 0.2500, 0.0000])

# -- expected implementation (according to Accuracy = `(TN + TP) / (N + P) = (128 + 105) / (128 + 7 + 105 + 2)`)


_target = torch.tensor([one_hot(torch.tensor(y), num_classes=7).tolist() for y in _target])
_preds = torch.tensor([one_hot(torch.argmax(p).cpu(), num_classes=7).tolist() for p) in _preds])

acc = Accuracy(task="multilabel", num_labels=NUM_CLASSES, labels=dataset.classes, average='weighted')
# tensor(0.9352)

acc = Accuracy(task="multilabel", num_labels=NUM_CLASSES, labels=dataset.classes, average=None)
# tensor([0.9628, 0.9256, 0.9917, 0.8719, 1.0000, 0.9339, 1.0000])
```

Of course the behaviour between sklearn and torchmetrics is constant, so this is probably me ...
Just for anyone else stumbling upon this issue :sweat_smile:
